### PR TITLE
CUSTCOM-179 PrimeFaces AutoSelect Event is Null within Custom Component

### PR DESCRIPTION
**Pull request by MeroRai**

### CODE REVIEW

File: api/src/main/java/javax/el/ELUtil.java [hunk 1]
```diff
@@ -222,13 +222,22 @@ static Method findMethod(Class<?> klass, String methodName, Class<?>[] paramType
      */
     static Object invokeMethod(ELContext context, Method method, Object base, Object[] params) {
 
-        Object[] parameters = buildParameters(context, method.getParameterTypes(), method.isVarArgs(), params);
         try {
+            int paramCount = params == null ? 0 : params.length;
+            if (paramCount == 0) {
+                for (Class ParameterType : method.getParameterTypes()) {
+                    if (ParameterType.getName().equals("javax.faces.event.AjaxBehaviorEvent")) {
```
Comment (dmatej) [inline api/src/main/java/javax/el/ELUtil.java:229]: I don't like it much, relying on class names is extremely fragile.
Also ... what it does ... 
IF we received null or zero parameters THEN look at the method signature, and IF ANY of those parameters is of exact class name execute the method with what we received.

IN ALL OTHER cases transform parameters to new parameter array and execute the method.

File: api/src/main/java/javax/el/ELUtil.java [hunk 2]
```diff
@@ -222,13 +222,22 @@ static Method findMethod(Class<?> klass, String methodName, Class<?>[] paramType
...
+                }
+            }
+            Object[] parameters = buildParameters(
+                    context, method.getParameterTypes(), method.isVarArgs(), params);
             return method.invoke(base, parameters);
         } catch (IllegalAccessException iae) {
             throw new ELException(iae);
         } catch (IllegalArgumentException iae) {
-            throw new ELException(iae);
+            throw new IllegalArgumentException(iae);
```
Comment (dmatej) [inline api/src/main/java/javax/el/ELUtil.java:240]: you can rethrow iae directly. But ... to me it looks like it was explicitly desired to throw ELException instead.
I think that in good code this catch block should not be there. But first we need to know why is it here.

Comment (dmatej) [reply api/src/main/java/javax/el/ELUtil.java:240]: I'm reviewing it for second time and I think the ElException is desired here, so it  should be reset back or explained at least.

File: impl/src/main/java/com/sun/el/util/ReflectionUtil.java [hunk 3]
```diff
@@ -146,15 +144,24 @@ public static PropertyDescriptor getPropertyDescriptor(Object base, Object prope
...
-    public static Object invokeMethod(ELContext context, Method m, Object base, Object[] params) {
+    public static Object invokeMethod(ELContext context, Method method, Object base, Object[] params) {
 
-        Object[] parameters = buildParameters(context, m.getParameterTypes(), m.isVarArgs(), params);
         try {
-            return m.invoke(base, parameters);
+            int paramCount = params == null ? 0 : params.length;
+            if (paramCount == 0) {
+                for (Class ParameterType : method.getParameterTypes()) {
+                    if (ParameterType.getName().equals("javax.faces.event.AjaxBehaviorEvent")) {
```
Comment (dmatej) [inline impl/src/main/java/com/sun/el/util/ReflectionUtil.java:153]: This is really ugly fragile hack. Bad code, antipattern.

File: impl/src/main/java/com/sun/el/util/ReflectionUtil.java [hunk 4]
```diff
@@ -146,15 +144,24 @@ public static PropertyDescriptor getPropertyDescriptor(Object base, Object prope
...
+                }
+            }
+            Object[] parameters = buildParameters(
+                    context, method.getParameterTypes(), method.isVarArgs(), params);
+            return method.invoke(base, parameters);
         } catch (IllegalAccessException iae) {
             throw new ELException(iae);
         } catch (IllegalArgumentException iae) {
-            throw new ELException(iae);
+            throw new IllegalArgumentException(iae);
```
Comment (dmatej) [inline impl/src/main/java/com/sun/el/util/ReflectionUtil.java:164]: Dtto, explain this change and delete catch block XOR leave original ELException here.

**Review by dmatej — CHANGES_REQUESTED** [review]

**Comment by MeroRai** [comment]
@dmatej I agree with your review. As I have explained [here](https://github.com/payara/Payara/pull/4592), this issue is caused by PrimeFaces. This is the only way I could find to fix it outside PrimeFaces, as it invokes without an event on its first attempt and only invokes with an event when an exception is thrown.

**Review by dmatej — APPROVED** [review]
1) The fix works with PrimeFaces 5.2 and Primefaces 8.0
2) The bug is in PrimeFaces, so this is only a temporary workaround on our side as it seems it does not break other things.

**Pull request merged**